make_abi_library.py

```diff
@@ -15,34 +12,41 @@
 STATS = {}
 
 
-def has_function(args):
+def has_function(seen=None, *args):
```

Comment:
Maybe rename this to:

```python
def _has_function(seen, *args):
  ...

def has_function(*args):
  global FUNCTION_CACHE
  has_func = False
  for ctype in args:
    if ctype not in FUNCTION_CACHE:
      FUNCTION_CACHE[ctype] = _has_function(set(), ctype)
    has_func = has_func or FUNCTION_CACHE[ctype]
  return has_func
```

Reply:
With this setup, you would no longer need to check if `seen is None`, and inside `_has_function`, you'd also call `_has_function` and you wouldn't assign to `FUNCTION_CACHE`.

Reply:
updated